packages/addon/lua/gelly/mods/blood-mod/init.lua

```diff
@@ -85,32 +155,44 @@ local function sprayBlood(damageType, attacker, position, force, damage)
 	density = density + damageMultiplier * damage
 
 	gellyx.emitters.Sphere({
-		center = position,
+		center = config.FromEntity and
+				 victim:GetPos() or
+				 position,
 		velocity = velocity,
 		radius = config.CubeSize,
 		density = density,
 		randomness = config.Randomness,
+		material = material,
 	})
 end
 
+local function convertTableToMaterial(table, material)
+	material.Roughness = 			  table.Roughness
+	material.IsSpecularTransmission = table.IsSpecularTransmission
+	material.RefractiveIndex = 		  table.RefractiveIndex
+	material.Absorption = 			  table.Absorption
+	material.DiffuseColor = 		  table.DiffuseColor
+end
+
 hook.Add(
 	"GellyXDamage",
 	"gelly.builtin.blood-mod",
-	function(_, attacker, position, force, damage, type)
+	function(victim, attacker, position, force, damage, type)
 		if
-			not attacker:IsPlayer()
-			and not attacker:IsNPC()
-			and attacker:GetClass() ~= "prop_ragdoll"
+			not victim:IsValid() or
+			(not victim:IsPlayer() and
+			not victim:IsNPC() and
+			not victim:IsRagdoll())
 		then
 			return
 		end
-		-- 8194 is the combination used in M9K for bullets
-		if type == 8194 then
-			type = 2
-		end
 
-		print(type)
-		sprayBlood(type, attacker, position, force, damage)
+		local bloodColor = victim:GetBloodColor() or BLOOD_COLOR_RED
+		local material = gellyx.presets.copyPresetMaterial("Blood")
+
+		convertTableToMaterial(BLOOD_COLOR_MATERIALS[bloodColor], material)
```

Comment:
nit: This entire statement can be factored out, you can just use `BLOOD_COLOR_MATERIALS[bloodColor]` as the material override now. I added a copy material function in case you want to handle any potential future changes of the base immutable blood preset in any future updates.